### PR TITLE
Allow setter/getter with a different type

### PR DESCRIPTION
**Pull request by dstepanov**

### CODE REVIEW

**Comment by dstepanov** [comment]
Initially, I also wanted to separate annotations into read/write, based on the accessors, but that doesn't really work well with Jackson. I still kept the API in AST.

**Review by graemerocher — APPROVED** [review]

File: core-processor/src/main/java/io/micronaut/inject/beans/visitor/BeanIntrospectionWriter.java [hunk 1]
```diff
@@ -419,32 +419,95 @@ private void buildStaticInit(ClassWriter classWriter) {
...
             beanPropertyData.name,
-            beanPropertyData.typedElement,
-            beanPropertyData.annotationMetadata,
-            beanPropertyData.typeArguments,
+            beanPropertyData.type,
             defaults,
             loadTypeMethods
         );
+
+        int typeLocal = -1;
```
Comment (yawkat) [inline core-processor/src/main/java/io/micronaut/inject/beans/visitor/BeanIntrospectionWriter.java:435]: Could use Supplier.memoized

Comment (yawkat) [reply core-processor/src/main/java/io/micronaut/inject/beans/visitor/BeanIntrospectionWriter.java:435]: meh probably not more readable though

File: inject-groovy-test/src/main/groovy/io/micronaut/ast/transform/test/AbstractBeanDefinitionSpec.groovy [hunk 2]
```diff
@@ -243,7 +243,7 @@ abstract class AbstractBeanDefinitionSpec extends Specification {
      *
      * @return the introspection if it is correct
      **/
-    protected BeanIntrospection buildBeanIntrospection(String className, @Language("groovy") String cls) {
+    protected BeanIntrospection buildBeanIntrospection(String className, String cls) {
```
Comment (yawkat) [inline inject-groovy-test/src/main/groovy/io/micronaut/ast/transform/test/AbstractBeanDefinitionSpec.groovy]: why this change?

**Review by yawkat — APPROVED** [review]
does this add significant size to the generated introspections?

File: core-processor/src/main/java/io/micronaut/inject/beans/visitor/BeanIntrospectionWriter.java (hunk 1 above)
Comment (dstepanov) [reply core-processor/src/main/java/io/micronaut/inject/beans/visitor/BeanIntrospectionWriter.java:435]: That would be more confusing; the invocation is contextual, and if a variable is not needed, we don't create it.

File: inject-groovy-test/src/main/groovy/io/micronaut/ast/transform/test/AbstractBeanDefinitionSpec.groovy (hunk 2 above)
Comment (dstepanov) [reply inject-groovy-test/src/main/groovy/io/micronaut/ast/transform/test/AbstractBeanDefinitionSpec.groovy]: Reverted

**Comment by dstepanov** [comment]
Not much, considering we generate an additional argument only if the setter/getter value is different (in most cases, it shouldn't), plus it needs to be enabled.
